### PR TITLE
Ekf2 replay

### PR DESCRIPTION
**Pull request by tumbili**
Replaces #3606

### CODE REVIEW

**Comment by tumbili** [comment]
Tested, works well!

**Comment by LorenzMeier** [comment]
Does that mean good to merge?

**Comment by tumbili** [comment]
@LorenzMeier Somebody should review it!

**Comment by LorenzMeier** [comment]
@tumbili Travis fail.

**Comment by LorenzMeier** [comment]
@tumbili Good to merge if my review determines its ok?

**Comment by tumbili** [comment]
@LorenzMeier Mainly have a look at the diff in the logging app. The rest is mostly the replay app itself which won't break anything else.

File: src/modules/sdlog2/sdlog2.c [hunk 1]
```diff
@@ -981,20 +1004,18 @@ int sdlog2_thread_main(int argc, char *argv[])
...
 	if (log_rate_ph != PARAM_INVALID) {
-		int32_t param_log_rate;
 		param_get(log_rate_ph, &param_log_rate);
 
 		if (param_log_rate > 0) {
 
 			/* we can't do more than ~ 500 Hz, even with a massive buffer */
-			if (param_log_rate > 500) {
-				param_log_rate = 500;
+			if (param_log_rate > 250) {
```
Comment (tumbili) [inline src/modules/sdlog2/sdlog2.c]: @LorenzMeier Is there a parameter I can get the sensor loop rate from?

Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: Not really. You would need to do the dt locally.

File: src/modules/ekf2/ekf2_main.cpp [hunk 2]
```diff
@@ -273,7 +284,13 @@ void Ekf2::task_main()
 	// initialise parameter cache
 	updateParams();
 
+	// initialize data structures outside of loop
+	// because they will else not always be
+	// properly populated
+	sensor_combined_s sensors = {};
```
Comment (LorenzMeier) [inline src/modules/ekf2/ekf2_main.cpp]: have you ensured the logging app does not exceed the stack size on Pixhawk?

File: src/modules/sdlog2/CMakeLists.txt [hunk 3]
```diff
@@ -31,7 +31,7 @@
 #
 ############################################################################
 if (${OS} STREQUAL "nuttx")
-	list(APPEND MODULE_CFLAGS -Wframe-larger-than=1500)
+	list(APPEND MODULE_CFLAGS -Wframe-larger-than=2500)
```
Comment (LorenzMeier) [inline src/modules/sdlog2/CMakeLists.txt]: This is not good. It should only be increased as needed and the app stack adjusted as well. This warning is there for a reason, and we're low on RAM!

File: src/modules/sdlog2/sdlog2.c [hunk 4]
```diff
@@ -1280,8 +1313,67 @@ int sdlog2_thread_main(int argc, char *argv[])
...
+
+	int poll_to_logging_factor = 1;
+
+	if (record_replay_log) {
+		subs.replay_sub = orb_subscribe(ORB_ID(ekf2_replay));
+		fds[0].fd = subs.replay_sub;
+		fds[0].events = POLLIN;
+		poll_to_logging_factor = 1;
+	} else {
+		subs.sensor_sub = orb_subscribe(ORB_ID(sensor_combined));
```
Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: Why can't we subscribe to sensor_combined instead?

File: src/modules/ekf2/ekf2_main.cpp (hunk 2 above)
Comment (tumbili) [inline src/modules/ekf2/ekf2_main.cpp]: I'll check this again but we should be fine.

File: src/modules/sdlog2/sdlog2.c (hunk 4 above)
Comment (tumbili) [inline src/modules/sdlog2/sdlog2.c]: I wasn't sure if we then know 100% that we log all replay messages. Couldn't it be that the app misses a sample then?

**Comment by tumbili** [comment]
@LorenzMeier I realized that the sdlog2 app used all of the stack. So I increased it such that it has a bit of free stack left. However, I don't understand this concept of frame size so maybe you can explain this to me.

**Comment by tumbili** [comment]
Added final changes. Will give this one more test flight in replay and non-replay mode. Should be good to go.

**Comment by tumbili** [comment]
@LorenzMeier This is good to go from my side. I was able to reduce the stack size and the frame size again. If someone could just quickly test that logging is working for non replay mode (EKF2_REC_RPL = 0) then it can go in.

**Comment by LorenzMeier** [comment]
Rebased and applied.

**Pull request closed without merging**